app/src/org/commcare/activities/LoginActivityUIController.java

```diff
@@ -441,13 +436,31 @@ private void setupLoginButton() {
         loginButton.setTextColor(getResources().getColor(R.color.cc_neutral_bg));
     }
 
-    protected void restoreLastUser() {
+    /**
+     *
+     * @return if a username was restored
+     */
+    protected boolean restoreLastUsername() {
         SharedPreferences prefs = CommCareApplication._().getCurrentApp().getAppPreferences();
-        String lastUser = prefs.getString(CommCarePreferences.LAST_LOGGED_IN_USER, null);
-        if (lastUser != null) {
-            username.setText(lastUser);
+
+        // First try to restore the last username that was entered, but NOT successfully logged in
+        // with (this value gets cleared upon a successful login so that that can take precedence)
+        String lastEnteredUsername = prefs.getString(LoginActivity.KEY_LAST_ENTERED_USERNAME, null);
```

Comment:
Note that the only reason I'm retaining the more complicated workflow of checking for BOTH the last entered username and then the last logged in username (as opposed to just allowing the last entered username to serve both purposes) is for the differential behavior of setting the focus placement (at the end of the username entry box for the last-entered username, under the assumption that they may still be in the middle of typing it, vs. in the password entry box for last logged in username). If this doesn't seem like a good enough reason for the extra complexity, I could remove it, but I do think it's a nicer UX this way